src/codegen/implicit-caller-location.md

```diff
@@ -0,0 +1,278 @@
+explained in terms of how it works in a `const` context.
+
+## Caller Location in `const`
+
+There are two main phases to returning the caller location in a const context: walking up the stack
+to find the right location and allocating a const value to return.
+
+### Finding the right `Location`
+
+In a const context we "walk up the stack" from where the intrinsic is invoked, stopping when we 
+reach the first function call in the stack which does *not* have the attribute. This walk is in
+[`InterpCx::find_closest_untracked_caller_location()`][const-find-closest].
+
+Starting at the bottom, we iterate up over stack [`Frame`][const-frame]s in the [InterpCx::stack],
+calling [`InstanceDef::requires_caller_location`][requires-location] on the 
+[`Instance`s from each `Frame`][frame-instance]. We stop once we find one that returns `false` and
+return the span of the *previous* frame which was the "topmost" tracked function.
+
+### Allocating a static `Location`
+
+Once we have a `Span` we need to allocate static memory for the `Location`, which is performed by 
+the [`TyCtxt::const_caller_location()`][const-location-query] query. Internally this calls 
+[`InterpCx::alloc_caller_location()`][alloc-location] and results in a unique
+[memory kind][location-memory-kind]. The SSA codegen backend is able to emit code for these same 
+values, and we use this code there as well..
+
+Once our `Location` has been allocated in static memory, our intrinsic returns a reference to it.
+
+## Generating code for `#[track_caller]` callees
+
+To generate efficient code for a tracked function and its callers we need to provide the same 
+behavior from the intrinsic's point of view without having a stack to walk up at runtime. We invert
+the approach: as we grow the stack down we pass an additional argument to calls of tracked functions
+rather than walking up the stack when the intrinsic is called. That additional argument can be
+returned wherever the caller location is queried.
+
+The argument we append is of type `&'static core::panic::Location<'staic>`. A reference was chosen
+to avoid unnecessary copying because a pointer is a third the size of 
+`std::mem::size_of::<core::panic::Location>() == 24` at time of writing.
+
+When generating a call to a function which is tracked, we pass the location argument the value of
+[`FunctionCx::get_caller_location`][fcx-get].
+
+If the calling function is tracked, `get_caller_location` returns the local in
+[`FunctionCx::caller_location`][fcx-location] which was populated by the current caller's caller.
+In these cases the intrinsic "returns" a reference which was actually provided in an argument to its
+caller.
+
+If the calling function is not tracked, `get_caller_location` allocates a `Location` static from
+the current `Span` and returns a reference to that.
+
+By chaining together the `caller_location` fields of multiple `FunctionCx`s as we grow the bottom of
```

Comment:
Nit: the term "chaining together" is confusing to me. Are we building a linked list in memory?

Reply:
Oh, I guess we are. OK, great.

Reply:
Not AFAIK, it's more of a conceptual chaining, every `#[track_caller]` function using their location parameter, to which a `#[track_caller]` caller would've passed in their own location parameter etc.

Reply:
`We achieve the same behavior as a loop starting from the bottom by passing a single `&Location` value through the `caller_location` fields of multiple `FunctionCx`s as we grow the stack downward."

?

Reply:
Actually just going to commit that and push, it's the last bit of feedback.